packages/eslint-plugin/src/rules/prefer-optional-chain.ts

```diff
@@ -1,669 +1,1521 @@
+        return NodeComparisonResult.Invalid;
+      }
 
-    /**
-     * Gets a normalized representation of the given MemberExpression
-     */
-    function getMemberExpressionText(node: TSESTree.MemberExpression): string {
-      let objectText: string;
+      return NodeComparisonResult.Equal;
+    }
 
-      // cases should match the list in ALLOWED_MEMBER_OBJECT_TYPES
-      switch (node.object.type) {
-        case AST_NODE_TYPES.MemberExpression:
-          objectText = getMemberExpressionText(node.object);
-          break;
+    case AST_NODE_TYPES.TemplateElement: {
+      const nodeBElement = nodeB as typeof nodeA;
+      if (nodeA.value.cooked === nodeBElement.value.cooked) {
+        return NodeComparisonResult.Equal;
+      }
+      return NodeComparisonResult.Invalid;
+    }
 
-        case AST_NODE_TYPES.CallExpression:
-        case AST_NODE_TYPES.Identifier:
-        case AST_NODE_TYPES.MetaProperty:
-        case AST_NODE_TYPES.ThisExpression:
-          objectText = getText(node.object);
-          break;
+    // these aren't actually valid expressions.
+    // https://github.com/typescript-eslint/typescript-eslint/blob/20d7caee35ab84ae6381fdf04338c9e2b9e2bc48/packages/ast-spec/src/unions/Expression.ts#L37-L43
+    case AST_NODE_TYPES.ArrayPattern:
+    case AST_NODE_TYPES.ObjectPattern:
+      /* istanbul ignore next */
+      return NodeComparisonResult.Invalid;
 
-        /* istanbul ignore next */
-        default:
-          return '';
-      }
+    // update expression returns a number and also changes the value each time - so it makes no sense to compare the chain
+    case AST_NODE_TYPES.UpdateExpression:
+      return NodeComparisonResult.Invalid;
 
-      let propertyText: string;
-      if (node.computed) {
-        // cases should match the list in ALLOWED_COMPUTED_PROP_TYPES
-        switch (node.property.type) {
-          case AST_NODE_TYPES.Identifier:
-            propertyText = getText(node.property);
-            break;
+    // yield returns the value passed to the `next` function, so it may not be the same each time - so it makes no sense to compare the chain
+    case AST_NODE_TYPES.YieldExpression:
+      return NodeComparisonResult.Invalid;
 
-          case AST_NODE_TYPES.Literal:
-          case AST_NODE_TYPES.TemplateLiteral:
-          case AST_NODE_TYPES.BinaryExpression:
-            propertyText = sourceCode.getText(node.property);
-            break;
+    // general-case automatic handling of nodes to save us implementing every
+    // single case by hand. This just iterates the visitor keys to recursively
+    // check the children.
+    //
+    // Any specific logic cases or short-circuits should be listed as separate
+    // cases so that they don't fall into this generic handling
+    default:
+      return compareByVisiting(nodeA, nodeB);
+  }
+}
+const COMPARE_NODES_CACHE = new WeakMap<
+  TSESTree.Node,
+  WeakMap<TSESTree.Node, NodeComparisonResult>
+>();
+function compareNodes(
```

Comment:
FWIW I found it a good bit easier to parse through this file when I moved `compareNodes` and `CompareNodesArgument` to a new file (along with the things only used by `compareNodes`).

Reply:
Yeah it'd definitely be a good idea to modularise this file so that it is a bit more broken up and isn't as massive.